demos/elastodynamics/elastodynamics-mwe.py

```diff
@@ -0,0 +1,233 @@
+
+def sigma(u):
+    return lmbda*div(u)*Identity(3) + 2*mu*epsilon(u)
+
+# Select whether or not the mass matrix will be lumped::
+lump_mass = True
+
+# Set time parameters::
+T                = 0.4
+dt               = 1e-3
+Nsteps_numerical = int(T/dt)
+t                = 0.
+step             = 0
+
+# We define Young's modulus and Poisson's ratio::
+E  = 50.0e6
+nu = 0.3
+
+# Next we calculate Lame's constants::
+lmbda   = E*nu/((1 + nu)*(1 - 2*nu))
+mu      = E/(2*(1 + nu))
+rho     = 1986.
+
+# The loading parameters are::
+amplitude = 40.0e3
+omega     = 50.
+
+# We create a rectangular column mesh where the number of elements in the
+# :math:`z`-direction is allowed to vary to demonstrate convergence::
+height = 20.
+width  = 1.
+nx     = 1
+ny     = 1
+nz     = 20
+mesh   = BoxMesh(nx, ny, nz, width, width, height)
+
+# Set up outputs::
+u_numerical    = np.zeros((Nsteps_numerical+2))
+time_numerical = np.linspace(0, T, Nsteps_numerical+2)
+
+# We first compute all surface areas of the mesh and then assign
+# the differential area element :math: `\Gamma_N`:: at the top surface
+ds      = Measure('ds', mesh)
+Gamma_N = ds(6)
+
+# We choose a degree 1 continuous vector function space, and set up
+# the function space and functions::
+W = VectorFunctionSpace(mesh, 'Lagrange', 1)
+u = TrialFunction(W)
+w = TestFunction(W)
+
+u_n   = Function(W, name='displacement')
+u_np1 = Function(W)
+u_nm1 = Function(W)
+
+# We next establish the Dirichlet boundary conditions::
+fixedSides  = Constant(0.)
+fixedBottom = as_vector((0.,0.,0.))
+
+fixedLeft_BC_x  = DirichletBC(W.sub(0), fixedSides, 1)
+fixedLeft_BC_y  = DirichletBC(W.sub(1), fixedSides, 1)
+fixedRight_BC_x = DirichletBC(W.sub(0), fixedSides, 2)
+fixedRight_BC_y = DirichletBC(W.sub(1), fixedSides, 2)
+fixedBack_BC_x  = DirichletBC(W.sub(0), fixedSides, 3)
+fixedBack_BC_y  = DirichletBC(W.sub(1), fixedSides, 3)
+fixedFront_BC_x = DirichletBC(W.sub(0), fixedSides, 4)
+fixedFront_BC_y = DirichletBC(W.sub(1), fixedSides, 4)
+fixedBottom_BC  = DirichletBC(W, fixedBottom, 5)
+
+bcSet = [fixedLeft_BC_x, fixedRight_BC_x, fixedBack_BC_x, fixedFront_BC_x, \
+         fixedLeft_BC_y, fixedRight_BC_y, fixedBack_BC_y, fixedFront_BC_y, \
+         fixedBottom_BC]
+
+# Next we define the traction vector which will be updated in the time loop::
+t = Constant(t)
+traction = as_vector([0, 0, -0.5*amplitude*(1 - cos(omega*t))])
+
+# Next we define the bilinear forms::
+a = dot(u, w)*dx
+L = (dt**2/rho) * (dot(traction, w)*Gamma_N - inner(sigma(u_n), epsilon(w))*dx) - dot(u_nm1 - 2*u_n, w)*dx
+
+print("Computing numerical solution...")
+
+if lump_mass:
+    
+    alump = replace(a, {u: as_vector([1, 1, 1])})
+    lumped_mass = assemble(alump)
+
+    # Time-stepping loop::
+    for step in range(1, Nsteps_numerical + 2):
+        # Update time step::
+        t.assign(t + dt)
+
+        assemble(L, tensor=u_np1, bcs=bcSet)
+
+        # Divide by lumped mass matrix::
+        u_np1 /= lumped_mass
```

Comment:
Thank you both for your feedback. @ksagiyam [spotted](https://github.com/firedrakeproject/firedrake/pull/2296#discussion_r765695973) that I had updated `u_nm1` after updating `u_n` in the lumped mass case which is what was throwing the error. I have anyways implemented @wence- 's suggestion that we write the form once. However, this block still gives an inaccurate solution, even after correcting the stress term to leave the method explicit. I tested with the consistent mass matrix which looks fine, so I believe I wrote the form correctly.

Reply:
I think you want to apply the bcs _after_ dividing by the lumped mass (I think it probably isn't the cause of the error in this case because the bcs are all zero) since otherwise you don't have the bc value on the boundary nodes but rather the bc value divided through by the lumped mass matrix. I'll try and find some time to run this demo to see what is going on though...